### DISCUSSION TITLE
AutoPath for object keys

### DISCUSSION THREAD

**Discussion by kbromwich** [millsp/ts-toolbelt#262] [Q&A]
The AutoPath function is quite cool, but it's not clear (to me) if it's possible to use/implement the AutoPath function such that it applies to the keys of an object.

For example, Mongo allows you to perform an update with the "$set" operator, such as:
```typescript
MyTable.update(id, { $set: {
  'data.subfield.value1': false,
  'data.subfield.value2': 2,
  'data.value': 'example',
} });
```

I'd like to add typing to a function that mimics that behaviour (and also to a function that will execute a Mongo update using the "$set" operator, but we'll stick with the mimic functionality example for now):
```typescript
const indexCreate = (obj: any, key: string) => {
  let val = obj[key];
  if (val === undefined || val === null) {
    val = {};
    obj[key] = val;
  }
  return val;
};
const updateObj = <T, K extends Record<string, any>>(obj: T, dottedKeyVals: K): T => {
  const result = Array.isArray(obj) ? [...obj] : { ...obj };
  Object.entries(dottedKeyVals).forEach(([dotPath, value]) => {
    const parts = dotPath.split('.');
    const penult = parts.slice(0, -1).reduce(indexCreate, result);
    penult[parts[parts.length - 1]] = value;
  });
  return result;
};

interface SampleInterface {
  a: {
    b: number;
  };
  c: string;
}
const inputObj: SampleInterface = { a: { b: 1}, c: 'str' };
const result = updateObj(inputObj, {
  'a.b': 2,
  'c': 'newStr',
});
```
In the above example, is it possible to use AutoPath (or something else) to type the keys (and values) of the `dottedKeyVals` object (and maybe the updated obj return type as well? Though I guess that should just be T)?

Maybe this asking a bit much, though. Maybe it's a "more trouble than it's worth" situation. 😅

**Comment by theoephraim** [comment]
Also looking to use dot path notation within object keys and arrays. Any guidance would be much appreciated!
